File: wk06/d2/crazy_bubble_sort.js
var readline = require('readline');
var reader = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});


var crazyBubbleSort = function(arr, sortCompletionCallback) {
    var sortPassCallback = function(swapsMade) {
        if (swapsMade) {
            performSortPass(arr, 0, false, sortPassCallback);
        } else {
            sortCompletionCallback(arr);
        }
    };
    performSortPass(arr, 0, false, sortPassCallback);
};

var performSortPass = function (arr, i, madeAnySwaps, runAnotherLoopCallback) {
    if (i < arr.length - 1) {
        askLessThan(arr[i], arr[i + 1], function (lessThan) {
            if (!lessThan) {
                var temp = arr[i];
                arr[i] = arr[i + 1];
                arr[i + 1] = temp;
                performSortPass(arr, i + 1, true, runAnotherLoopCallback);
            } else {
                performSortPass(arr, i + 1, madeAnySwaps, runAnotherLoopCallback);
            }
        });
    } else {
        runAnotherLoopCallback(madeAnySwaps);
    } 
};

var askLessThan = function (el1, el2, swappingCodeCallback) {
    reader.question("Is " + el1 + " less than " + el2 + "?", 
        function (yesOrNo) {
            if (yesOrNo === 'yes') {
                swappingCodeCallback(true);
            } else {
                swappingCodeCallback(false);
            }
        }
    );
};

crazyBubbleSort([3, 2, 1], function (arr) { console.log(arr); });

File: wk06/d2/crazy_bubble_sort.rb
def crazyBubbleSort(arr)
  ####### looping construct, aka the sortPassCallback
  swapsMade = true
  while swapsMade do
    # ******** actual swapping, the performSortPass method
    swapsMade = false
    arr.each_with_index do |el, i| 
      # $$$$$$$ ask user to compare
      lessThan = gets.chomp
      # $$$$$$$ ask user to compare
      if lessThan # this has to be a callback to the user input function
        arr[i], arr[i + 1] = arr[i + 1], arr[i]
        swapsMade = true
      end
    end
    # ******** actual swapping, the performSortPass method
  end
  ####### looping construct, aka the sortPassCallback
  
  # this is where the sortCompletionCallback goes
  puts arr
end